lib/cfa/augeas_parser/writer.rb

```diff
@@ -0,0 +1,333 @@
+      # no entry in tree, then do not insert position, which means, that
+      # following setting of value appends it to the end.
+      #
+      # @param located_entry [LocatedEntry] entry to insert
+      # @return [String] return string to where value should be written. Can
+      #   contain path expressions.
+      #   See https://github.com/hercules-team/augeas/wiki/Path-expressions
+      def insert_entry(located_entry)
+        # entries with add not exist yet
+        preceding = located_entry.preceding_existing
+        prefix = located_entry.prefix
+        if preceding
+          insert_after(preceding, located_entry)
+        # entries with remove is already removed, otherwise find previously
+        elsif located_entry.any_following?
+          aug.insert(prefix + "/*[1]", located_entry.key, true)
+          aug.match(prefix + "/*[1]").first
+        else
+          "#{prefix}/#{located_entry.key}"
+        end
+      end
+
+      # Insert key after preceding.
+      # @see insert_entry
+      # @param preceding [LocatedEntry] entry after which should be written new
+      #   entry
+      # @param located_entry [LocatedEntry] to insert
+      # @return [String] return string to where value should be written.
+      def insert_after(preceding, located_entry)
+        aug.insert(preceding.path, located_entry.key, false)
+        paths = aug.match(located_entry.prefix + "/*")
+        paths_index = paths.index(preceding.path) + 1
+        paths[paths_index]
+      end
+    end
+
+    attr_reader :aug
+
+    # Do modification according to operation defined in AugeasEntry
+    # @param tree [CFA::AugeasTree] tree where entry lives
+    # @param entry [AugeasElement] entry to process
+    # @param last_valid_entry_path [String, nil] path of last valid entry
+    #   written or nil if there is no such entry
+    # @param key [String] valid augeas key for entry. Important for collections
+    #   which have key without its index, but augeas do not allow it.
+    # @param path [String] whole path where to write entry in augeas
+    def process_operation(located_entry)
+      case located_entry.entry[:operation]
+      when :add, nil then @lazy_operations.add(located_entry)
+      when :remove then @lazy_operations.remove(located_entry)
+      when :modify then modify_entry(located_entry)
+      when :keep then recurse_write(located_entry)
+      else raise "invalid :operation in #{located_entry.inspect}"
+      end
+    end
+
+    # writes value of entry to path and if it have sub-tree then call {write}
+    # on it
+    # @param path [String] path where to write
+    # @param entry [AugeasElement] entry to write
+    def modify_entry(located_entry)
+      value = located_entry.entry_value
+      aug.set(located_entry.path, value)
+      report_error { aug.set(located_entry.path, value) }
+      recurse_write(located_entry)
+    end
+
+    # calls write on entry if entry have sub-tree
```

Comment:
ok, it in fact call it always, but if there is no sub-tree, then write is called on empty tree, so do nothing.